eis_toolkit/raster_processing/reclassify_raster.py

```diff
@@ -0,0 +1,511 @@
+    if number_of_classes <= 0:
+        raise InvalidParameterValueException("The number of classes must be at least one")
+
+    out_image, out_meta = _raster_with_geometrical_intervals(raster, number_of_classes, bands)
+
+    return out_image, out_meta
+
+
+def _raster_with_standard_deviation(  # type: ignore[no-any-unimported]
+    raster: rasterio.io.DatasetReader,
+    number_of_intervals: int,
+    bands: Optional[Sequence[int]],
+) -> Tuple[Sequence[np.ndarray], dict]:
+
+    out_image = []
+    out_meta = raster.meta.copy()
+
+    band_statistics = []
+    if bands is not None:
+        for band in bands:
+            stddev = np.nanstd(band)
+            mean = np.nanmean(band)
+            band_statistics.append((mean, stddev))
+    else:
+        for band in range(1, raster.count + 1):
+            stddev = np.nanstd(band)
+            mean = np.nanmean(band)
+            band_statistics.append((mean, stddev))
+
+    for band, (mean, std) in enumerate(band_statistics):
+        data_array = raster.read(band + 1)
+        interval_size = 2 * std / number_of_intervals
+
+        classified = np.empty_like(data_array)
+
+        below_mean = data_array < (mean - std)
+        above_mean = data_array > (mean + std)
+
+        classified[below_mean] = -number_of_intervals
+        classified[above_mean] = number_of_intervals
+
+        in_between = ~below_mean & ~above_mean
+        interval = ((data_array - (mean - std)) / interval_size).astype(int)
+        classified[in_between] = interval[in_between] - number_of_intervals // 2
+
+        out_image.append(classified)
+
+    return out_image, out_meta
+
+
+@beartype
+def raster_with_standard_deviation(  # type: ignore[no-any-unimported]
+    raster: rasterio.io.DatasetReader,
+    number_of_intervals: int,
+    bands: Optional[Sequence[int]],
+) -> Tuple[Sequence[np.ndarray], dict]:
+    """Classify raster with standard deviation.
+
+    If bands are not given, all bands are used for classification.
+
+    Args:
+        raster: Raster to be classified.
+        number_of_intervals: The number of intervals.
+        bands: Selected bands from multiband raster. Indexing begins from one. Defaults to None.
+
+    Returns:
+        Raster classified with standard deviation and metadata.
+
+    Raises:
+        InvalidParameterValueException: Bands is not a list, bands is not a list of integers,
+        bands exceeds the number of bands in the raster.
+    """
+    if bands is not None:
+        if not isinstance(bands, list):
+            raise InvalidParameterValueException("Expected bands parameter to be a list")
+        elif not all(isinstance(band, int) for band in bands):
+            raise InvalidParameterValueException("Expected bands to be a list of integers")
+        elif len(bands) > raster.count:
+            raise InvalidParameterValueException("The number of bands given exceeds the number of raster's bands")
```

Comment:
These checks are repeated across all the public functions. They could be refactored into a function that is then called to check the inputs to increase code reuse. E.g.

~~~python
def _check_reclassify_raster_inputs(bands):
    if not isinstance(bands, list):
        raise InvalidParameterValueException("Expected bands parameter to be a list")
    elif not all(isinstance(band, int) for band in bands):
        raise InvalidParameterValueException("Expected bands to be a list of integers")
    elif len(bands) > raster.count:
        ...
~~~